File: packages/agentx-native/src/api/universal.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum FinishReason {
    Stop,
    Length,
    ToolCall,
    ContentFilter,
    Error,
    Unknown,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ContentBlock {
    pub r#type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub arguments: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<Vec<ContentBlock>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_error: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thinking: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum UniversalItem {
    Message {
        role: Role,
        #[serde(skip_serializing_if = "Option::is_none")]
        id: Option<String>,
        content: Vec<ContentBlock>,
    },
    ToolCall {
        id: String,
        name: String,
        arguments: Value,
    },
    ToolResult {
        tool_call_id: String,
        content: Vec<ContentBlock>,
        #[serde(skip_serializing_if = "Option::is_none")]
        is_error: Option<bool>,
    },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct UniversalTool {
    pub name: String,
    pub description: Option<String>,
    pub input_schema: Option<Value>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ToolChoice {
    pub r#type: String,
    pub name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct GenerationConfig {
    pub temperature: Option<f64>,
    pub top_p: Option<f64>,
    pub max_output_tokens: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct UniversalRequest {
    pub id: Option<String>,
    pub model: Option<String>,
    #[serde(default)]
    pub instructions: Vec<ContentBlock>,
    #[serde(default)]
    pub input: Vec<UniversalItem>,
    #[serde(default)]
    pub tools: Vec<UniversalTool>,
    #[serde(default)]
    pub tool_choice: Option<ToolChoice>,
    #[serde(default)]
    pub stream: bool,
    #[serde(default)]
    pub generation: GenerationConfig,
}

File: packages/agent-core/agent/api/universal.py
from enum import Enum
from typing import List, Optional, Dict, Any, Union
from pydantic import BaseModel, Field

class Role(str, Enum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"

class FinishReason(str, Enum):
    STOP = "stop"
    LENGTH = "length"
    TOOL_CALL = "tool_call"
    CONTENT_FILTER = "content_filter"
    ERROR = "error"
    UNKNOWN = "unknown"

class ContentBlock(BaseModel):
    type: str
    text: Optional[str] = None
    id: Optional[str] = None
    name: Optional[str] = None
    arguments: Optional[Any] = None
    tool_call_id: Optional[str] = None
    content: Optional[List['ContentBlock']] = None
    is_error: Optional[bool] = None
    thinking: Optional[str] = None
    signature: Optional[str] = None
    data: Optional[str] = None
    media_type: Optional[str] = None
    url: Optional[str] = None

class UniversalItem(BaseModel):
    type: str
    role: Optional[Role] = None
    id: Optional[str] = None
    content: Optional[List[ContentBlock]] = None
    name: Optional[str] = None
    arguments: Optional[Any] = None
    tool_call_id: Optional[str] = None
    is_error: Optional[bool] = None
    text: Optional[str] = None
    encrypted: Optional[str] = None
    raw: Optional[Any] = None

class UniversalTool(BaseModel):
    name: str
    description: Optional[str] = None
    input_schema: Optional[Dict[str, Any]] = None

class ToolChoice(BaseModel):
    type: str
    name: Optional[str] = None

class GenerationConfig(BaseModel):
    temperature: Optional[float] = None
    top_p: Optional[float] = None
    max_output_tokens: Optional[int] = None

class ReasoningConfig(BaseModel):
    effort: Optional[str] = None
    budget_tokens: Optional[int] = None
    visible: Optional[bool] = None

class UniversalRequest(BaseModel):
    id: Optional[str] = None
    model: Optional[str] = None
    instructions: List[ContentBlock] = Field(default_factory=list)
    input: List[UniversalItem] = Field(default_factory=list)
    tools: List[UniversalTool] = Field(default_factory=list)
    tool_choice: Optional[ToolChoice] = None
    stream: bool = False
    generation: GenerationConfig = Field(default_factory=GenerationConfig)
    reasoning: Optional[ReasoningConfig] = None
